src/storage/storage.cpp

```diff
@@ -94,32 +126,63 @@ int Storage::Run()
     }
 #endif
 
-    try
+    auto regions_layout = getRegionsLayout(barriers);
+    const SharedDataType layout_region = regions_layout.old_layout_region;
+    const SharedDataType data_region = regions_layout.old_data_region;
+
+    if (max_wait > 0)
+    {
+        util::SimpleLogger().Write() << "Waiting for " << max_wait
+                                     << " second for all queries on the old dataset to finish:";
+    }
+    else
+    {
+        util::SimpleLogger().Write() << "Waiting for all queries on the old dataset to finish:";
+    }
+
+    boost::interprocess::scoped_lock<boost::interprocess::named_sharable_mutex> regions_lock(
+        regions_layout.old_regions_mutex, boost::interprocess::defer_lock);
+
+    if (max_wait > 0)
     {
-        boost::interprocess::scoped_lock<boost::interprocess::named_mutex> pending_lock(
-            barrier.pending_update_mutex);
+        if (!regions_lock.timed_lock(boost::posix_time::microsec_clock::universal_time() +
+                                     boost::posix_time::seconds(max_wait)))
+        {
+            util::SimpleLogger().Write(logWARNING) << "Queries did not finish in " << max_wait
```

Comment:
Add a comment here about the consequences to the downstream reader processes if this happens.